controllers/topology/reconcile_state.go

```diff
@@ -114,27 +124,35 @@ func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, s *scope.
 // ControlPlane objects should never change (only the content of the objects can change).
 func (r *ClusterReconciler) reconcileCluster(ctx context.Context, s *scope.Scope) error {
 	log := ctrl.LoggerFrom(ctx)
+	log.V(4).Info("Reconciling the Cluster object")
 
 	// Check differences between current and desired state, and eventually patch the current object.
 	patchHelper, err := mergepatch.NewHelper(s.Current.Cluster, s.Desired.Cluster, r.Client)
 	if err != nil {
 		return errors.Wrapf(err, "failed to create patch helper for %s/%s", s.Current.Cluster.GroupVersionKind(), s.Current.Cluster.Name)
 	}
-	if patchHelper.HasChanges() {
-		log.Info("Updating Cluster")
-		if err := patchHelper.Patch(ctx); err != nil {
-			return errors.Wrapf(err, "failed to patch %s/%s", s.Current.Cluster.GroupVersionKind(), s.Current.Cluster.Name)
-		}
+	if !patchHelper.HasChanges() {
+		log.V(3).Info(fmt.Sprintf("No changes for %s/%s", s.Current.Cluster.GroupVersionKind(), s.Current.Cluster.Name))
+		return nil
+	}
+
+	log.Info(fmt.Sprintf("Patching Cluster %s/%s", s.Current.Cluster.GroupVersionKind(), s.Current.Cluster.Name))
+	if err := patchHelper.Patch(ctx); err != nil {
+		return errors.Wrapf(err, "failed to patch %s/%s", s.Current.Cluster.GroupVersionKind(), s.Current.Cluster.Name)
 	}
 	return nil
 }
 
 // reconcileMachineDeployments reconciles the desired state of the MachineDeployment objects.
 func (r *ClusterReconciler) reconcileMachineDeployments(ctx context.Context, s *scope.Scope) error {
+	log := ctrl.LoggerFrom(ctx)
+
 	diff := calculateMachineDeploymentDiff(s.Current.MachineDeployments, s.Desired.MachineDeployments)
 
 	// Create MachineDeployments.
 	for _, mdTopologyName := range diff.toCreate {
+		log.V(4).Info(fmt.Sprintf("Reconciling MachineDeployment with topology-name %s (create)", mdTopologyName))
```

Comment:
Couldn't we create a new context/logger here with pre-defined values based on the MachineDeployment we're reconciling and passing it down?

Reply:
I can explore this but,  given the problem with values described above, I'm not sure this will prevent us to do fmt down in the call stack